matchbox_socket/src/webrtc_socket/socket.rs

```diff
@@ -211,6 +211,17 @@ pub struct WebRtcSocket {
 }
 
 impl WebRtcSocket {
+    /// Creates a new builder for a connection to a given room with a given number of
+    /// re-connection attempts.
+    ///
+    /// You must add at least one channel with [`WebRtcSocketBuilder::add_channel`],
+    /// [`WebRtcSocketBuilder::add_reliable_channel`], or
+    /// [`WebRtcSocketBuilder::add_unreliable_channel`] before you can build the
+    /// [`WebRtcSocket`]
+    pub fn builder(room_url: impl ToString, attempts: Option<u16>) -> WebRtcSocketBuilder {
```

Comment:
room_url and attempts should be something applied at the builder level, e.g.

```rs
WebRtcSocket::builder()
    .room_url(IpAddr::v4(..))
    .port(3536)
    .attempts(3)
    ...

Reply:
One thing i don't like about this, is that it makes it easier to use the api incorrectly. You could end up trying to build a socket without a signaling server, or without any channels.

For the common use case, I think i prefer if we keep some simple, non-builder methods as an alternative, I.e

```rust
let socket = WebRtcSocket::connect_reliable("wss://example.com"):
```

Also, I think it's good if the builder constructor forces you to provide a signalling server in its new method, as all other fields can have reasonable defaults.

Reply:
The canonical builder pattern supposed to have arguments in `new` if they cannot be given sensible defaults, I have given attempts a default of `Some(3)` and removed it from the `new` function. See: [rust-unofficial pattern docs](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html)

Reply:
I think the builder pattern is a wonderful addition, personally. I'd absolutely approve.

Hm, I didn't know that @garryod . That's fine then. I think Some(3) is a reasonable default for attempts then.